enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,609 @@
+- an error in thrown from the IPAM CNI to the user if the NAD does **not**
+  allow for persistent IPs
+
+#### Stopping a Virtual Machine
+```mermaid
+sequenceDiagram
+actor user
+
+user ->> KubeVirt: stopVM(vm=vmA)
+
+KubeVirt ->> Kubernetes: deletePod(owner=vmA)
+Kubernetes -->> KubeVirt: OK
+
+KubeVirt -->> user: OK
+
+note over OVN-K: notices a pod was deleted
+OVN-K ->> OVN-K: ips = teardownOVNEntities()
+
+alt is deleted pod orphaned ?
+OVN-K ->> Kubernetes:   deletePersistentIPs(ips)
+Kubernetes -->> OVN-K: OK
+end
+```
+
+The main difference from a traditional IPAM CNI response to a pod delete is the
+`PersistentIP` allocation will not be deleted if the CNI DEL arguments
+feature any owner indication. Thus, on the VM scenario, the `PersistentIP`s
+will **not** be deleted.
+
+#### Starting a (previously stopped) Virtual Machine
+```mermaid
+sequenceDiagram
+
+actor user
+
+user ->> KubeVirt: startVM(vm=vmA)
+
+KubeVirt ->> Kubernetes: createPod(name=<generated name>, owner=vmA)
+
+
+note over OVN-K: notices a new pod was added
+OVN-K ->> Kubernetes: getPersistentIPs(ownerLabel = <...>)
+Kubernetes -->> OVN-K: persistentIPs
+OVN-K ->> Kubernetes: result(..., ips = persistentIPs)
+
+Kubernetes -->> KubeVirt: OK
+
+KubeVirt -->> user: OK
+```
+
+When creating the pod, the IPAM CNI finds existing `PersistentIP`s for the
+workload. It will thus use those already reserved allocations, instead of
+generating brand new allocations for the pod where the encapsulating object
+will run.
+
+#### Removing a Virtual Machine
+```mermaid
+sequenceDiagram
+
+actor user
+
+user ->> KubeVirt: stopVM(vm=vmA)
+
+KubeVirt ->> Kubernetes: deletePod(owner=vmA)
+
+note over OVN-K: notices a pod was deleted
+OVN-K ->> OVN-K: ips = teardownOVNEntities()
+
+Kubernetes -->> KubeVirt: OK
+
+KubeVirt -->> user: OK
+
+loop every allocation whose "owner" was deleted:
+    Kubernetes ->> Kubernetes: garbageCollect(allocation)
+end
+```
+
+This scenario, from an IPAM CNI perspective is exactly like deleting a pod:
+since the `PersistentIP` is owned by a Kubernetes entity we give up on
+manipulating the datastore for it, and delegate to Kubernetes to garbage
+collect them once the encapsulating object is removed.
+
+### API Extensions
+
+- Adds a new `persistentips` CRD. The CNI plugin will provision these to "block"
+the IP allocation during the encapsulating object lifecycle.
+  - It must be "owned" by the pod's encapsulating object - thus allowing
+  Kubernetes to GC it once they're gone.
+  - It must be "retrievable" by the plugin (i.e. so the "old" VM gets the same
+  IP when in runs in a "new" pod). For that, we label the CR w/ information
+  which will be available to the CNI plugin.
+
+This is the proposed spec of the CRD:
+```yaml
+apiVersion: "persistentip.cni.cncf.io"
+kind: PersistentIP
+metadata:
+  name: my-shiny-vm-persistent-ip
+  namespace: ns1
+  labels:
+    persistentips.cni.cncf.io/owner: VirtualMachine.vm-a.a0790345-4e84-4257-837a-e3d762d191ab
```

Comment:
This must be split into name, namespace, reference. Otherwise you are at risk of hitting the limit of label length ["must be 63 characters or less (can be empty),"](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set)